src/core/stream/period/get_adaptation_switch_strategy.ts

```diff
@@ -64,22 +64,66 @@ export default function getAdaptationSwitchStrategy(
     return { type: "continue", value: undefined };
   }
 
-  // remove from that intersection what we know to be the right Adaptation
-  const adaptationInBuffer = getBufferedRangesFromAdaptation(queuedSourceBuffer,
-                                                             period,
-                                                             adaptation);
-  const { currentTime } = clockTick;
+  queuedSourceBuffer.synchronizeInventory();
+  const inventory = queuedSourceBuffer.getInventory();
+
+  // Continue if we have no other Adaptation buffered in the current Period
+  if (!inventory.some(buf => buf.infos.period.id === period.id &&
+                             buf.infos.adaptation.id !== adaptation.id))
+  {
+    return { type: "continue", value: undefined };
+  }
+
+  /** Data already in the right Adaptation */
+  const adaptationInBuffer =
+    getBufferedRangesFromAdaptation(inventory, period, adaptation);
+
+  /**
+   * Data different from the wanted Adaptation in the Period's range.
+   * /!\ Could contain some data at the end of the previous Period or at the
+   * beginning of the next one.
+   */
+  const unwantedRange = excludeFromRanges(intersection, adaptationInBuffer);
+
+  if (unwantedRange.length === 0) {
+    return { type: "continue", value: undefined };
+  }
+
+  const { currentTime } = playbackInfo;
   if (adaptation.type === "video" &&
-      clockTick.readyState > 1 &&
+      // We're playing the current Period
       isTimeInRange({ start, end }, currentTime) &&
-      (isTimeInRanges(bufferedRanges, currentTime) &&
-       !isTimeInRanges(adaptationInBuffer, currentTime)))
+      // There is data for the current position
+      playbackInfo.readyState > 1 &&
+      // We're not playing the current wanted video Adaptation
+      !isTimeInRanges(adaptationInBuffer, currentTime))
   {
     return { type: "needs-reload", value: undefined };
   }
 
-  const unwantedData = excludeFromRanges(intersection, adaptationInBuffer);
+  // From here, clean-up data from the previous Adaptation, if one
+
+  const rangesToExclude = [];
+
+  // First, we don't want to accidentally remove some segments from the previous
+  // Period (which overlap a little with this one)
+
+  /** Last segment before one for the current period. */
+  const lastSegmentBefore = getLastSegmentBeforePeriod(inventory, period);
+  if (lastSegmentBefore !== null &&
+    (lastSegmentBefore.bufferedEnd === undefined ||
+      period.start - lastSegmentBefore.bufferedEnd < 1)) // Close to Period's start
```

Comment:
Do you thing it would be good to add a config value for the "1" second to exclude ?

Reply:
I hesitated. It becomes very hard to name and could look out of place in the config